python/src/equistore/operations/to.py

```diff
@@ -0,0 +1,179 @@
+        )
+    if not isinstance(backend, str):
+        raise TypeError("`backend` should be passed as a `str`")
+    if backend == "torch":
+        # Check args
+        if not isinstance(requires_grad, bool):
+            raise TypeError("`requires_grad` should be passed as a `bool`")
+        if not isinstance(dtype, torch.dtype):
+            raise TypeError("`dtype` should be passed as a `torch.dtype`")
+        if not isinstance(device, torch.device):
+            raise TypeError("`device` should be passed as a `torch.device`")
+        return _to_torch_block(block, requires_grad, dtype, device)
+    elif backend == "numpy":  # numpy
+        # TensorBlock
+        return _to_numpy_block(block)
+    # `backend` is not "numpy" or "torch"
+    raise ValueError(f"backend ``{backend}`` not supported")
+
+
+def _to_torch_block(
+    block: TensorBlock,
+    requires_grad: bool,
+    dtype: torch.dtype,
+    device: torch.device,
+) -> TensorBlock:
+    """
+    Creates a new :py:class:`TensorBlock` where block values are PyTorch
+    :py:class:`torch.tensor` objects. Assumes the block values are already as a
+    type that is convertible to a :py:class:`torch.tensor`, such as a numpy
+    array or RustNDArray. The resulting torch tensor dtypes are enforced as
+    :py:class:`torch.float64`.
+
+    :param block: input :py:class:`TensorBlock`, with block values as ndarrays.
+    :param requires_grad: bool, whether or not to torch's autograd should record
+        operations on this tensor.
+    :param dtype: ``torch.dtype``, the base dtype of the resulting torch tensor,
+        i.e. ``torch.float64``.
+    :param device: ``torch.device``, the device on which the resulting torch
+        tensor should be stored, i.e. ``torch.device("cpu")``.
+
+    :return: a :py:class:`TensorBlock` whose values tensor is now of type
+        :py:class:`torch.tensor`.
+    """
+    # Create new block, with the values tensor converted to a torch tensor.
+    new_block = TensorBlock(
+        values=torch.tensor(block.values, requires_grad=requires_grad, dtype=dtype).to(
+            device.type
+        ),
+        samples=block.samples,
+        components=block.components,
+        properties=block.properties,
+    )
+
+    # Add gradients to each block, again where the values tensor are torch
+    # tensors.
+    for parameter, gradient in block.gradients():
+        new_block.add_gradient(
+            parameter,
+            torch.tensor(gradient.data, requires_grad=requires_grad, dtype=dtype).to(
+                device.type
+            ),
+            gradient.samples,
+            gradient.components,
+        )
+    return new_block
+
+
+def _to_numpy_block(block: TensorBlock) -> TensorBlock:
+    """
+    Takes a TensorBlock object whose values are torch.tensor objects and
+    converts them to numpy arrays of dtype np.float64. Returns a new TensorBlock
+    object.
+    """
+    if isinstance(block.values, np.ndarray):
+        return block.copy()
```

Comment:
What memory considerations should we make here @Luthaf @frostedoyster ? Do we want to provide the option to release the previous tensor from memory? If so, should the default behaviour be safe (return a copy) or unsafe?

See discussion in https://github.com/lab-cosmo/equistore/pull/248